identity/src/main/java/com/android/identity/CredentialData.java

```diff
@@ -1299,8 +1302,24 @@ Collection<X509Certificate> getAuthKeysNeedingCertification() {
                             aliasForAuthKey,
                             KeyProperties.PURPOSE_SIGN | KeyProperties.PURPOSE_VERIFY)
                             .setDigests(KeyProperties.DIGEST_SHA256, KeyProperties.DIGEST_SHA512);
-                    kpg.initialize(builder.build());
-                    kpg.generateKeyPair();
+
+                    KeyPair keyPair = null;
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                        builder.setIsStrongBoxBacked(true);
+                        try {
+                            kpg.initialize(builder.build());
+                            keyPair = kpg.generateKeyPair();
+                            Log.i(TAG, "AuthKey created in StrongBox");
+                        } catch (StrongBoxUnavailableException e) {
```

Comment:
Using an exceptions for something that's actually expected on a whole class of devices is odd. It also makes the code more complicated, as we have to retry on failure (rather than doing the right thing in the first place).

It's more straightforward to check for the feature: `packageManager.hasSystemFeature(PackageManager.FEATURE_STRONGBOX_KEYSTORE`.

That reduces much of this function to a single conditional
`if (isStrongBoxSupported) { setIsStrongBoxBacked(isStrongBoxSupported) }`

I believe this also means you can get rid of the sdk version check, as well.

Reply:
Fair enough, doing this. Still need the SDK version check, I'm afraid.